subprojects/workers/src/main/java/org/gradle/workers/internal/DefaultWorkerExecutor.java

```diff
@@ -0,0 +1,260 @@
+        } else {
+            sharedPackagesBuilder.add(visibleClass.getPackage().getName());
+        }
+    }
+
+    private static class RunnableWorkFuture implements ListenableFuture<Void> {
+        private final String description;
+        private final ListenableFuture<DefaultWorkResult> delegate;
+        private final AtomicBoolean userHandled = new AtomicBoolean(false);
+
+        public RunnableWorkFuture(String description, ListenableFuture<DefaultWorkResult> delegate) {
+            this.description = description;
+            this.delegate = delegate;
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            return delegate.cancel(mayInterruptIfRunning);
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return delegate.isCancelled();
+        }
+
+        @Override
+        public boolean isDone() {
+            return delegate.isDone();
+        }
+
+        @Override
+        public Void get() throws InterruptedException, ExecutionException {
+            userHandled.set(true);
+            DefaultWorkResult result = delegate.get();
+            throwIfNotSuccess(result);
+            return null;
+        }
+
+        @Override
+        public Void get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+            userHandled.set(true);
+            DefaultWorkResult result = delegate.get(timeout, unit);
+            throwIfNotSuccess(result);
+            return null;
+        }
+
+        void getIfNotHandled() throws InterruptedException, ExecutionException {
+            if (!userHandled.get()) {
+                // if the user has not already taken responsibility for the error, fail if the
+                // work fails
+                get();
+            } else {
+                try {
+                    delegate.get();
+                } catch (Throwable t) {
+                    // The user has already taken responsibility for the error, so we just make sure the
+                    // work has finished and ignore any errors.
+                }
+            }
+        }
+
+        @Override
+        public void addListener(Runnable listener, Executor executor) {
+            delegate.addListener(listener, executor);
+        }
+
+        private void throwIfNotSuccess(DefaultWorkResult result) throws ExecutionException {
+            if (!result.isSuccess()) {
+                throw new WorkExecutionException(description, result.getException());
+            }
+        }
+    }
+
+    @Contextual
+    private static class WorkExecutionException extends ExecutionException {
+        public WorkExecutionException(String description) {
+            this(description, null);
+        }
+
+        public WorkExecutionException(String description, Throwable cause) {
+            super("A failure occurred while executing " + description + ": " + cause.getMessage(), cause);
```

Comment:
We should avoid this pattern of including the cause's message in the contextual exception's message